read-fonts/src/tables/truetype/mem.rs

```diff
@@ -0,0 +1,192 @@
+            alloc_slice(buf, glyph.max_other_points)?
+        } else {
+            (Default::default(), buf)
+        };
+        // Don't allocate any delta buffers if we don't have variations
+        let (deltas, iup_buffer, composite_deltas, buf) = if glyph.has_variations {
+            let (deltas, buf) = alloc_slice(buf, glyph.max_simple_points)?;
+            let (iup_buffer, buf) = alloc_slice(buf, glyph.max_simple_points)?;
+            let (composite_deltas, buf) = alloc_slice(buf, glyph.max_component_delta_stack)?;
+            (deltas, iup_buffer, composite_deltas, buf)
+        } else {
+            (
+                Default::default(),
+                Default::default(),
+                Default::default(),
+                buf,
+            )
+        };
+        let (contours, buf) = alloc_slice(buf, glyph.contours)?;
+        let (flags, _) = alloc_slice(buf, glyph.points)?;
+        Some(Self {
+            unscaled,
+            scaled,
+            original_scaled,
+            contours,
+            flags,
+            deltas,
+            iup_buffer,
+            composite_deltas,
+        })
+    }
+}
+
+/// Trait that defines which types may be used as element types for
+/// `alloc_slice`.
+///
+/// # Safety
+/// This must only be implemented for types that contain no internal padding
+/// and for which all bit patterns are valid.
+unsafe trait TransmuteElement: Sized + Copy {}
+
+unsafe impl TransmuteElement for u16 {}
+unsafe impl TransmuteElement for i32 {}
+unsafe impl TransmuteElement for Fixed {}
+unsafe impl TransmuteElement for F26Dot6 {}
+unsafe impl TransmuteElement for PointFlags {}
+unsafe impl<T> TransmuteElement for Point<T> where T: TransmuteElement {}
+
+/// Allocates a mutable slice of `T` of the given length from the specified
+/// buffer.
+///
+/// Returns the allocated slice and the remainder of the buffer.
+fn alloc_slice<T: TransmuteElement>(buf: &mut [u8], len: usize) -> Option<(&mut [T], &mut [u8])> {
+    if len == 0 {
+        return Some((Default::default(), buf));
+    }
+    // 1) Ensure we slice the buffer at a position that is properly aligned
+    // for T.
+    let base_ptr = buf.as_ptr() as usize;
+    let aligned_ptr = align_up(base_ptr, align_of::<T>());
+    let aligned_offset = aligned_ptr - base_ptr;
+    let aligned_buf = buf.get_mut(aligned_offset..)?;
+    // 2) Ensure we have enough space in the buffer to allocate our slice.
+    let len_in_bytes = len * size_of::<T>();
+    if len_in_bytes > aligned_buf.len() {
+        return None;
+    }
+    let (slice_buf, rest) = aligned_buf.split_at_mut(len_in_bytes);
+    // SAFETY: Alignment and size requirements have been checked in 1) and
+    // 2) above, respectively.
+    // Element types are limited by the `TransmuteElement` trait which
+    // defines requirements for transmutation and is private to this module.
+    let slice = unsafe { std::slice::from_raw_parts_mut(slice_buf.as_mut_ptr() as *mut T, len) };
```

Comment:
I believe this is fine on the basis that:
- this function takes a borrow on buf
- thus the returned slice borrow keeps the original borrow from the caller
- the returned borrows do not overlap, as provided by split_at_mut() and that this slice's length in bytes was used to do the splitting
- getting the borrow as a pointer then making a new borrow is fine, the original one is invalidated but was thrown away, slice_buf is not usable thereafter.

Reply:
Thanks for the additional analysis. This all looks correct to me as well.